clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -268,6 +268,8 @@ private void completeFutureAndFireCallbacks(long baseOffset, long logAppendTime,
             // A newly created batch can always host the first message.
             if (!batch.tryAppendForSplit(record.timestamp(), record.key(), record.value(), record.headers(), thunk)) {
                 batches.add(batch);
+                // Close the batch for record appends to free up resources like compression buffers.
+                batch.closeForRecordAppends();
```

Comment:
```batch.closeForRecordAppends``` does NOT release the byte array hosted by ```MemoryRecordsBuilder```
```java
    public void closeForRecordAppends() {
        if (appendStream != CLOSED_STREAM) {
            try {
                appendStream.close();
            } catch (IOException e) {
                throw new KafkaException(e);
            } finally {
                appendStream = CLOSED_STREAM;
            }
        }
    }
```

```appendStream``` is a wrap of ```bufferStream```. Not sure whether calling ```batch.closeForRecordAppends()``` can resolve the OOM or not.

```java
this.appendStream = new DataOutputStream(compressionType.wrapForOutput(this.bufferStream, magic));
```

Reply:
@chia7712 Thanks for reporting the issue. But I'd like to understand this problem a bit more. Supposedly the split batches will be closed when they are drained out of the `RecordAccumulator`.
https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java#L608

and `ProducerBatch#close()` calls `ProducerBatch#closeForRecordAppends()`.

Do you have a heap dump when the OOM happens so we can take a further look?

Reply:
@becketqin In my case, the OOM is caused by negative estimatedCompressionRatio. I have committed #8285 to fix it.